git/filter_process_scanner.go

```diff
@@ -0,0 +1,199 @@
+// Package git contains various commands that shell out to git
+// NOTE: Subject to change, do not rely on this package from outside git-lfs source
+package git
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/github/git-lfs/errors"
+	"github.com/rubyist/tracerx"
+)
+
+// FilterProcessScanner provides a scanner-like interface capable of
+// initializing the filter process with the Git parent, and scanning for
+// requests across the protocol.
+//
+// Reading a request (and errors) is as follows:
+//
+//     s := NewFilterProcessScanner(os.Stdin, os.Stderr)
+//     for s.Scan() {
+//             req := s.Request()
+//     	       // ...
+//     }
+//
+//     if err := s.Err(); err != nil {
+//             // ...
+//     }
+type FilterProcessScanner struct {
+	// pl is the *pktline instance used to read and write packets back and
+	// forth between Git.
+	pl *pktline
+
+	// req is a temporary variable used to hold the value accessible by the
+	// `Request()` function. It is cleared at the beginning of each `Scan()`
+	// invocation, and written to at the end of each `Scan()` invocation.
+	req *Request
```

Comment:
Ha, good nit. I did not separate these by a newline because `req` and `err` are both used in _very_ similar ways when calling the `Scan()` (and internal `readRequest()`) function.